### ISSUE TITLE
Remote-control Spotify using the Connect protocol, especially the Windows client?

### ISSUE THREAD

**Issue by CoDEmanX** [sashahilton00/spotify-connect-resources#2]
Hi, I wonder if any of the research on Spotify Connect allows controlling a Spotify client yet.

I'm really not interested in receiving and decoding audio data with a custom Spotify connect client, but only to perform actions like volume control and track queuing - essentially what I can do with my smartphone, Spotify's Android app and a premium subscription (for Spotify Connect).

It's very handy to be able to remote control the Windows client on my laptop, or my Spotify Connect-enabled speaker, using my phone. But I'd like to have these abilities independent of an official Spotify client (e.g. let a little Windows application query what songs I play on any of my devices and create some sort of custom last.fm to track what I hear the most; my speaker does unfortunately not support last.fm natively).

Any pointers are much appreciated!

**Comment by plietar** [comment]
By using my [librespot](https://github.com/plietar/librespot) library it could be done.
It currently only has an incomplete receiver side, but see https://github.com/plietar/librespot/blob/master/src/spirc.rs#L130 for the type of messages received, https://github.com/plietar/librespot/blob/master/src/spirc.rs#L175 on how to send some.

The basic spirc (Spotify Connect) protocol works by sending out a kMessageTypeHello first, and every receiver responds with a kMessageTypeNotify. They also send a broadcasted notify on startup and on status updates.

The protocol is fairly straightforward, there's a load, play, pause, seek, ... message
Unfortunately queueing is weirdly designed, and I'm not 100% sure how it all works

Let me know if you have any other queries on the protocol

**Comment by CoDEmanX** [comment]
Hey plietar! Thanks for the info. Unfortunately, I don't know any Rust, so the code is quite hard to follow. But I think I understand the basic principle.

So what you're saying is that spirc is a very simple message protocol, that can control a Spotify Connect enabled device. But there are only a few commands with no payload (i.e. no message body that contains a URI to play) if I understand the [DivideOverflow article](http://divideoverflow.com/2014/08/reversing-spotify-connect/#comment-1886018503) correctly - or do some of the mentioned commands actually come with a payload, e.g. `load`?

Is there any encryption / auth involved in spirc communication? I captured some network communication with Wireshark and snooped around the packets, but didn't see many readable messages (actually just something that contained a substring `SpotUpd` and what I think is the announcement broadcast of Spotify Connect devices: `_spotify-connect._tcp.local` - is that zeroconf?)

Is spirc transmitted and received in the local network only, or is there a three-way communication with a Spotify server (since Connect is a premium service)? Also, what is Mercury (`hm://`) and how does it relate to spirc?

Queuing isn't as important to me as playing a track immediately, and I have a feeling that Spotify got it a bit wrong with queues... you can't reorder / remove tracks from queue in the Android app, and it's not playing the queued tracks on my Spotify Connect speaker lately :disappointed: 

By incomplete receiver side, do you mean there are features like queuing missing, but it's possible to receive and decode audio streams (on linux)?

Thanks for bearing with me!

**Comment by plietar** [comment]
Spirc is a simple protocol running on top of spotify's more complex hermes/mercury, which itself runs on an encryption & authentication layer.

The protocol described in the divideoverflow article is as far as I know only some debug log sent to the spotify servers and serves no real purpose. The spirc messages (and almost everything else) are encoded using protobuf. The schema is described here : https://github.com/plietar/librespot/blob/master/protocol/proto/spirc.proto

Almost nothing is done on the local network, only discovery, using zeroconf, and sending a authentication token to the speaker which allows it to connect on your account.

Everything else goes through the server. Mercury is used as a PUB/SUB mechanism, where every message sent by a client is sent to every client connected on the same account. Clients must filter the messages themselves.

The receiver supports playing audio, along with basic commands, play pause next previous ... but queuing is missing

Writing a client shouldn't be too complicated, and I could really use one to control spotify without leaving my terminal :smile:  I might give it a go sometime soon.
I realize now that picking Rust might have been a mistake as it keeps people out from contributing.
OTOH, Rust is a great language so I invite anyone who has some time to spare to learn it :smile:

**Comment by CoDEmanX** [comment]
Rust looks like a pretty cool language for sure. A C++ implementation would attract more contributors I guess, but I'm not good with C++ either, so doesn't really matter to me ;) I would prefer a Node.js implementation in JavaScript, but not an expert with that either.

I get the concept of how Connect works (thanks again!), and also wonder how queuing can work. Is it possible that the Spotify server pushes queued songs to the client, and queuing them is handled outside of spirc (some other bits of hermes/mercury)?

Regarding APPKEY: what did you write to obtain one? Do they actually read what you enter in that form? I assume hardware vendors receive special app keys to be built into their Connect devices. Would be much easier to have one of those for the community, but they might compare it with the client info/behavior to verify that it's the device they issued the key for...

I'll try to setup a Rust development environment and read some tutorials, maybe we can code this together.

**Comment by sashahilton00** [comment]
I too would also love a node.js version, but it would still require some sort of native code for Hermes/mercury and audio decryption? I just put music app in the appkey box, and it worked. The appkey is auto generated, so I wouldn't worry too much. Also, there is a gentlemans agreement between this group and Spotify in the sense that normal appkeys given by spotify will work in this project, and in return we don't use the keys we extract from vendor firmware. Though if that changes, then obviously the vendor keys are on the table...

**Comment by sashahilton00** [comment]
Also, of you want to look for how the queue works, you might find out how by having a look at the JS on play.spotify.com. Part of me wonders if they don't just use the playlist manager for a specially named playlist like userxxxxx+queue and just render it differently in the app as the play queue, though there is no evidence for this, except that it would be perfectly possible to do.

**Comment by crsmoro** [comment]
i dont know if it helps..
one time because of a bug in spotify app on windows.. it showed uri instead of the names, so i discovered few things..
one of the thing is that what @sashahilton00 said is what happened.. it created a temporary playlist when i have manually added a song..

**Comment by CoDEmanX** [comment]
Standard TCP and UDP communication [is possible](https://nodejs.org/api/net.html) [out-of-the-box](https://nodejs.org/api/dgram.html) with Node.js as far as I can tell. Raw sockets require [additional modules](https://github.com/stephenwvickers/node-raw-socket), which are commonly written in C/C++ and compiled using node-gyp (which requires a viable build env).

Binary data can be handled using the [Buffer class](https://nodejs.org/api/buffer.html). There is a  [crypto](https://nodejs.org/api/crypto.html) module and another for [TLS/SSL](https://nodejs.org/api/tls.html), not sure if they are sufficient for the Hermes/mercury encryption.

As for audio, do you mean the aforementioned overall stream encryption, or the encoding of the audio data (MP3/AAC/...)? I'm sure there are some node modules that can decode such audio streams. And there's [node-speaker](https://github.com/TooTallNate/node-speaker), which can apparently play PCM data on your computer speakers. The author uses it in his [Node.js spotify client](https://github.com/TooTallNate/node-spotify-web), which uses the Spotify web protocol. We should have a closer look at this one, although communication is probably quite different and it can't control Connect devices. In fact, I managed to hear different tracks on my smartphone (using the Android client) and my laptop (using the web player) simultaneously...

As a coder, I would absolutely implement the queue as another playlist, because that's the most straightforward approach. The Web API can create and edit playlists, but [they claim it can't access the current queue](https://github.com/spotify/web-api/issues/15) (also see linked issues). So it doesn't seem to be implemented this way, or they block access to it out of security concerns...

**Comment by sashahilton00** [comment]
Sounds like it could be possible to do it in node.js then, i suspect there would be a fair bit of hacking stuff together, but node.js would be great. Regarding the audio, IIRC the audio is downloaded from Spotify CDN, and the files are 128 AES encrypted. Also on the crypto note, shannon will be an interesting thing to implement efficiently in node. Probably best to use node-gyp for that

**Comment by CoDEmanX** [comment]
Hm shannon as in entropy? There's already a module for Nodejs: https://github.com/aschuch/node-entropy

**Comment by sashahilton00** [comment]
Shannon as in the shannon stream cipher. https://eprint.iacr.org/2007/044.pdf as for rust, hopefully we can reuse some of the code, but it's all just greek to me ;)

**Comment by plietar** [comment]
Queue is NOT implemented as a playlist. It is entirely a Spirc concept. Spotify servers don't interact with spirc at all, other than forwarding messages between client.
You can witness that by the fact that if you close all clients and start them again, the play queue is lost.
If one client is active, the others can quit & join and the queue will be kept.

Shannon is implemented in C with a small Rust binding https://github.com/plietar/rust-shannon
I took the implementation from despotify. I haven't found it (or another implementation) anywhere non spotify related.
It is extremely likely that the official client (at least the desktop) uses the same implementation. The generated assembly code matches, and I was able to substitute certain functions with my own recompiled function which dump the unencrypted contents, and run the client like that. Extremely useful for examining the protocol.

The audio can either be downloaded over the Shannon encrypted TCP connection (this is what libspotify_embedded, despotify and librespot do) or over HTTP from spotify's CDN. My guess is they are moving to the CDN to reduce their load. I haven't fully figured out how the url to the CDN is retrieved, and what tokens need to be given. In any case, the files are encrypted using AES128 in CTR mode with a constant IV and a different key for each file. See https://github.com/plietar/librespot/blob/master/src/audio_decrypt.rs.
I've got very limited crypto knowledge, but I've always been told constant IVs are really bad. Maybe using different keys makes it OK.
The keys are always retrieved through the encrypted TCP connection.

Regarding app keys, AFAIK there's no such "gentlemans agreement" with Spotify. I've refused to store any app key (or shared library) on my repos, but there are some on this one.
Anyway, you can just get one from Spotify's developer website. It's the same ones libspotify uses

**Comment by sashahilton00** [comment]
Regarding the gentlemans agreement, i have spoken with an old colleague who has gone to work at Spotify. Long story short, they have enough on their plate there not to be too worried about an open source version of their connect implementation, so long as it isn't facilitating illegal (offline) downloading off Spotify, and we are not infringing the trademarks.

**Comment by CoDEmanX** [comment]
The paper @sashahilton00 referred to claimed, that there would be a C reference implementation available at [http://www.qualcomm.com.au/Shannon.html](https://web.archive.org/web/20081010141647/http://www.qualcomm.com.au/Shannon.html), but seems they never had one there. Maybe that's why Spotify chose it, because it's hard to find working code to decrypt Shannon? Or because it can run on basically every platform, and might perform well because of it's (comparatively) simplicity... they want you to listen to the music after all, not protect it from ever being deciphered ;)

Regarding illegal downloading: if any of the available custom clients would write the audio frames to disk instead of sending them to a speaker, wouldn't that already be a stream ripper? Such a tool would obviously violate the ToS. And if you shared the downloaded files, it would be copyright infringement for sure. But there's hardly a reason why one would do that as paying premium user anyway. I can see some legitimate use cases still, like if you want to use a DJ software to play some music at a private party and that software has no built-in Spotify integration.

@plietar: Do you know if the audio data is identical on both TCP and HTTP connections? The audio quality options seem to differ between the Android and the Windows client...

I got my own app key from Spotify, it was just the form that scared me a bit. As long as it's that simple to obtain, there's no reason not to use an own key.

Just tried to compile shn.c to JavaScript with Emscripten and it didn't print any errors. The optimized JS code is 154KB. Not sure if it's actually working.

**Comment by sashahilton00** [comment]
@CoDEmanX regarding the difficulty in finding a working implementation, I think that is exactly why Spotify chose it. Look @ slide 26-29 http://www.slideshare.net/mobile/ricardovice/spotify-p2p-music-streaming

**Comment by CoDEmanX** [comment]
Ouch... I managed to compile the dependencies of node-spotify-web after some trial and error (Node v0.10.41 + NPM v3.5.1 + a [node-speaker fork](https://github.com/RTK/node-speaker) + MSVC 2013 work for me on Windows 7) and it took me like 3 lines of code to turn it into a stream ripper. I consider this legal research, but I'm worried for Spotify that people might abuse it, if it's that easy to achieve. It doesn't seem to use the high-quality version of the MP3s, not sure if this is just another tiny change in spotify-web...

On the other hand, I'm surprised how well spotify-web works! It's kinda fun to play songs from the command line :smile:

**Comment by sashahilton00** [comment]
Obviously anything can be turned into a stream ripper, the point is, we are not setting out to create an application that is designed to do this. If someone goes off and creates a fork/modifies the code, then thats their problem. And lets be honest, it's not like anyone is going to bother when soundcloud downloader and youtube2mp3 exist

**Comment by CoDEmanX** [comment]
Turns out, they only serve MP3 160kbit/s for the web player. Maybe they don't offer 320kbit/s because of the potential piracy. The former is okay to listen to, but not very attractive to download.

But back to topic: I'm still interested in Connect, mainly remote control. Decoding audio is a bonus challenge. node-spotify-web proves Node.js being capable of being a Spotify client, although several dependencies (native modules) don't work with v0.11 and above. Some might be replaceable by pure JS implementations, alternative modules or more up-to-date forks. I think the main challenge is to handle the encryption and the queue magic though.

**Comment by CoDEmanX** [comment]
@plietar: I keep getting the following error if I `cargo build` on Windows:

```bash
C:\Users\CoDEmanX\.cargo\registry\src\github.com-88ac128001ac3a9a\mod_path-0.1.5\l
ib.rs:11:5: 11:28 error: unresolved import `rustc::plugin::Registry`. Could not
find `plugin` in `rustc` [E0432]
C:\Users\CoDEmanX\.cargo\registry\src\github.com-88ac128001ac3a9a\mod_path-0.1.5\l
ib.rs:11 use rustc::plugin::Registry;

             ^~~~~~~~~~~~~~~~~~~~~~~
```
Checked the Rust docs and everything looks okay to me. Any idea? I'm using Rust 1.7.0 nightly 64bit.

There's also a warning, which will become a hard error in a future version:

```bash
C:\Users\CoDEmanX\.cargo\registry\src\github.com-88ac128001ac3a9a\protobuf-1.0.4\s
rc/lib\codegen.rs:166:13: 166:31 warning: `RustType::Bool` does not name a tuple
 variant or a tuple struct [E0164]
C:\Users\CoDEmanX\.cargo\registry\src\github.com-88ac128001ac3a9a\protobuf-1.0.4\s
rc/lib\codegen.rs:166             RustType::Bool(..)                       => "f
alse".to_string(),

                                  ^~~~~~~~~~~~~~~~~~
```


> over HTTP from spotify's CDN (that's what the desktop client does)

So there's another HTTP interface? The web player endpoint serves 160kbit/s MP3s only, but the desktop client can be configured to stream high quality audio (320kbit/s MP3 I suppose).

Regarding Shannon: Is there an initial packet that contains the key to decode any further communication over the TCP connection?

**Comment by joerg-krause** [comment]
I run into the same issue: https://github.com/plietar/librespot/issues/11

**Comment by sashahilton00** [comment]
@CoDEmanX regarding the web interface and the 160kbps limit in the web interface, it is hardcoded into the Spotify Javascript SDK, and can be found by searching for <code>this.bitrate=160</code>. In case it is helpful, typing <code>DebuggerJS.on();</code> into the Spotify web console will also allow you to debug Spotify Messages.

**Comment by CoDEmanX** [comment]
Interesting finding, so some clients might use the RTMP protocol. I think the web interface is a dead end though, mainly because of the limited bitrate.

What I find quite interesting in the Connect device discovery: full clients (Windows, Android) react to the spotify mDNS question by also sending the same question, whereas thin clients (e.g. Connect speakers) also answer this question and offer a hostname, port and URL path for a JSON endpoint (and also a device name). Full clients do not broadcast their device name via mDNS. I suspect they are actually retrieved from the Spotify servers (they are permanently logged in to the user account, so the server knows them).

Regarding queuing: TrackRef is a repeated field in spirc, and every TrackRef structure comes with a "queued" property. So it stands to reason, that the queue is held and exchanged as part of the state. 
https://github.com/plietar/librespot/blob/master/protocol/proto/spirc.proto#L89
But I assume if it's this obvious, that @plietar would have figured it out already...

Regarding encryption: Do the official clients really use static values for the Diffie-Hellman key exchange calculations, or is this just a simplification?
https://github.com/plietar/librespot/blob/master/src/keys.rs#L10

**Comment by plietar** [comment]
The shannon keys are derived from the shared DH secret, along with the contents of the two initial unencrypted packets. See https://github.com/plietar/librespot/blob/master/src/keys.rs#L63

mDNS is only really used for credentials exchange. Since the user has to log in when using full clients it is not required, where as it is harder to type in you password into a speaker. Once logging in is done, everything goes through spotify servers.

A TrackRef is "queued" if it was added to the playing queue manually, instead of being part of the current playlist / album.

For DH, they always use the same group, but each side generates a random private/public key pair which is used to compute the shared secret. This is standard DH practice, and most HTTPS servers use the same group. As shown by recent paper, it may help an attacker as he can precompute stuff, but nothing to worry about : https://weakdh.org/imperfect-forward-secrecy-ccs15.pdf

You may be able to use spirc over the web interface. It gives access to mercury, so you could subscribe to the spirc channel and send/receive commands on there.

You could extremely easily make librespot a stream ripper, but I won't publish that.

**Comment by CoDEmanX** [comment]
Thanks for all the info again!

Well, I think the Shannon encryption is a showstopper for Node.js, because there is no native module available and I don't know how to write these (NAN, anyone?). As for the JavaScript port of the C code generated with Emscripten, I don't know how to test if it's actually working.

There's apparently another crypto suite available (`CRYPTO_SUITE_RC4_SHA1_HMAC`), but that is neither supported my Node natively. However, it does come with the building blocks for this if I understand correctly. But would that solely change how the stream is encrypted, or also the way Spotify communicates with the client?

**Comment by hongkongkiwi** [comment]
Whats to stop us using shannon in Node.js by using the shn.c file that the golang implementation uses?

**Comment by Fornoth** [comment]
On the topic of remote control, @badfortrains made this: 
https://github.com/badfortrains/spotcontrol/tree/master/examples/spot-chrome

**Comment by plietar** [comment]
@hongkongkiwi shannon is really not the issue IMO, you still have to implement the rest of the protocol, which is quite a bit of code. So far only @badfortrains did it in Go.
It would probably take less time to learn Rust and reuse librespot for your needs.

**Comment by badfortrains** [comment]
Yeah I would second @plietar's comment.   I would image the more painful part of doing a node implementation would be all the byte array manipulation used in other parts of the code.  Haven't really played with TypedArrays in node, but it seems like bit operations aren't the easiest to do in JS.  Also, rust is a cool language!  My main motivation for porting part of librespot over to Go was to learn more about rust and go.

That being said, if you're really set on node you could definitely write a wrapper for shn.c.  It's a little more involved than in Go or Rust but definitely doable.  
Another option would be using [gopherjs](https://github.com/gopherjs/gopherjs) (translates go code to javascript).  I used it to translate the entire spotcontrol code into javascript (which includes a go implementation of Shannon - itself converted from C via [c2go](https://godoc.org/rsc.io/c2go)) You could either just translate the shannon part, or use the whole thing and bridge some of the public methods to javascript.  If you did the later though you'd probably end up having to write some go code since the spotcontrol functionality is pretty limited.
